### PR TITLE
Bin-pack big VMs across clusters

### PR DESCRIPTION
**Pull request by leust**
Bin-pack big VMs across clusters

hana_* flavors cannot move automatically through DRS on spawning
and thus we need any VM going to a CUSTOM_HANA_EXCLUSIVE_HOST
cluster to be placed by Nova by filling the fullest nodes first.

Clusters now report memory_mb_max_unit representing the free
memory of the most free host in the cluster.

HANAMemoryMaxUnitFilter uses memory_mb_max_unit to filter out
hosts which don't have enough memory_mb_max_unit to fit the VM.

HANABinPackWeigher stacks hana VMs across the clusters.
It can be controlled via the configuration property
[filter_scheduler] hana_binpack_weight_multiplier where positive
values increase the stacking.

Change-Id: Ia1cc1e59b391bd83741e72b7e9462c810f2be2d0

### CODE REVIEW

File: nova/tests/unit/scheduler/weights/test_hana_binpack.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        # host4 should win:
+        weighed_host = self._get_weighed_host(hostinfo_list,
+                                              self._hana_spec_obj)
+        self.assertEqual('host4', weighed_host.obj.host)
+
+    def test_multiplier_zero(self):
+        self.flags(hana_binpack_weight_multiplier=0.0,
+                   group='filter_scheduler')
+        hostinfo_list = self._get_all_hosts()
+        # host4 should win:
```
Comment (joker-at-work) [inline nova/tests/unit/scheduler/weights/test_hana_binpack.py:73]: Is this a leftover from copy&paste from a different test?

**Review by joker-at-work — COMMENTED** [review]
We should write into the commit-message why we only do it for hana_* flavors. Imho, it's the assumption that hana_* flavors cannot move automatically through DRS on spawning and thus we need anything going to a `CUSTOM_HANA_EXCLUSIVE_HOST` cluster filtered by this filter.

We could add a test for a negative multiplier. This should make `host2` win, I guess.

File: nova/scheduler/weights/hana_binpack.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        return -1 * utils.get_weight_multiplier(
+            host_state, 'hana_binpack_weight_multiplier',
+            CONF.filter_scheduler.hana_binpack_weight_multiplier)
+
+    def _weigh_object(self, host_state, weight_properties):
+        """We want stacking to be the default."""
+        trait = "trait:%s" % BIGVM_EXCLUSIVE_TRAIT
+        extra_specs = weight_properties.flavor.extra_specs or {}
+        if extra_specs.get(trait) == "required":
+            return host_state.free_ram_mb
```
[the remark this answers is not in the document]
Comment (joker-at-work) [reply nova/scheduler/weights/hana_binpack.py:43]: We can overwrite `RamWeigher` by giving our weigher more weight, though.

File: nova/tests/unit/scheduler/filters/test_hana_memory_max_unit_filter.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        extra_specs = {'trait:CUSTOM_HANA_EXCLUSIVE_HOST': 'forbidden'}
+
+        flavor = objects.Flavor(memory_mb=3072,
+                                extra_specs=extra_specs)
+
+        spec_obj = objects.RequestSpec(
+            context=mock.sentinel.ctx,
+            flavor=flavor)
+
+        self.assertTrue(self.filt_cls.host_passes(host, spec_obj))
```
Comment (joker-at-work) [inline nova/tests/unit/scheduler/filters/test_hana_memory_max_unit_filter.py:72]: Since we're unit-testing, it might be worth to test the behavior when we pass in a flavor without `trait:CUSTOM_HANA_EXCLUSIVE_HOST`.

File: nova/scheduler/weights/hana_binpack.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    def _weigh_object(self, host_state, weight_properties):
+        """We want stacking to be the default."""
+        memory_mb_max_unit = utils.get_memory_mb_max_unit(host_state)
+        if memory_mb_max_unit is None:
+            return 0
+
+        trait = f"trait:{BIGVM_EXCLUSIVE_TRAIT}"
+        extra_specs = weight_properties.flavor.extra_specs
+        if extra_specs.get(trait) == "required":
+            return memory_mb_max_unit
```
Comment (fwiesel) [inline nova/scheduler/weights/hana_binpack.py:51]: I am not sure we want to do this.
That means, whenever I have a Big VM, regardless of how big it is, I will put it in the vcenter with the host with the most RAM. A.k.a. Worst-Fit allocation.

That leads to more fragmentation, and only makes sense, if we expect that VM to increase in size (in place).

Better is Random-Fit (i.e. disabling this weigher), or Best-Fit (i.e. the difference).

Comment (fwiesel) [reply nova/scheduler/weights/hana_binpack.py:51]: Ah, sorry.  I might have missed the *minus*.

**Review by fwiesel — APPROVED** [review]

**Review by Scsabiii — APPROVED** [review]
